### PR TITLE
Refactor Offerings

### PR DESCRIPTION
**Pull request by mistermoe**
# Summary
This PR is another iteration in our binary search towards finding the right package structure. Initial structure was too nested. 2nd iteration is far too flat and a bit unwieldy with 18,473 `With*` option functions (keenly pointed out by @phoebe-lew)

image

# Changes
put offerings related stuff into an `offerings` package. we can follow suit for each resource and message type. Generalized stuff like `Message`, `Resource`, `Sign`, `Verify` can stay in the `tbdex` package. 

Name collisions for consumers can be handled if relevant by aliasing the import e.g. 
```go
import tbdexoffering "github.com/TBD54566975/tbdex-go/tbdex/offering"
```

---

Refactored offering creation and removed 100+ lines of duplicated code. `Offering` is the chonkiest resource in tbDEX and creating one with endless `With*` Options was really hard to do. Switched over to something a bit simpler.

The refactor allows offerings to be created in several ways using the same underlying code:

#### Split out
incase the payin and payout fields are huge

```go
usdWirePayin := offering.NewPayin("USD", []offering.PayinMethod{
    offering.NewPayinMethod("WIRE"),
})

usdStoredBalancePayout := offering.NewPayout("USDC", []offering.PayoutMethod{
    offering.NewPayoutMethod("STORED_BALANCE", 20*time.Minute),
})

offering.Create(usdWirePayin, usdStoredBalancePayout, "1.0")
```

#### YOLO Structs
```go
offering.Create(
    offering.PayinDetails{
        CurrencyCode: "USD",
        Methods:      []offering.PayinMethod{
            offering.NewPayinMethod("WIRE"),
            offering.NewPayinMethod("SQUAREPAY"),
        },
    },
    offering.PayoutDetails{
        CurrencyCode: "USDC",
        Methods:      []offering.PayoutMethod{
            offering.NewPayoutMethod("STORED_BALANCE", 20*time.Minute),
        },
    },
    "1",
    offering.RequiredClaims(pexv2.PresentationDefinition{}),
)
```

---

Added Missing `RequiredClaims` field to `Offering`

> [!IMPORTANT]
> [RequiredClaims](https://github.com/TBD54566975/tbdex-go/blob/refactor/tbdex/offering/offering.go#L31) is a pointer to `pexv2.PresentationDefinition` in order to prevent marshaling/unmarshaling from setting it to its zero-state. Ideally would use @alecthomas 's [`optional.Option`](https://github.com/alecthomas/types/blob/master/optional/option.go) but it marshals to `null` for `nil` pointers which borks what the json schema expects. not necessarily happy with using a pointer but couldn't think of a better solution atm.

---

Added strongly typed representation for [OfferingID](https://github.com/TBD54566975/tbdex-go/blob/refactor/tbdex/offering/offering.go#L75-L81)

### CODE REVIEW

File: tbdex/offering/create.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+		description: fmt.Sprintf("%s for %s", payout.CurrencyCode, payin.CurrencyCode),
+		protocol:    "1.0",
+	}
+
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	if len(payin.Methods) == 0 {
+		return Offering{}, errors.New("1 payin method is required")
```
Comment (KendallWeihe) [inline tbdex/offering/create.go]:
```suggestion
		return Offering{}, errors.New("At least 1 payin method is required")
```

File: tbdex/offering/create.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	if len(payout.Methods) == 0 {
+		return Offering{}, errors.New("1 payout method is required")
```
Comment (KendallWeihe) [inline tbdex/offering/create.go]:
```suggestion
		return Offering{}, errors.New("At least 1 payout method is required")
```

**Comment by mistermoe** [comment]
@KendallWeihe error messages have to start with lowercase letter

**Comment by KendallWeihe** [comment]
> [RequiredClaims](https://github.com/TBD54566975/tbdex-go/blob/refactor/tbdex/offering/offering.go#L31) is a pointer to pexv2.PresentationDefinition in order to prevent marshaling/unmarshaling from setting it to its zero-state.

Not sure I fully understand, but could you write a json unmarshal/marshal override function for the `offering.Data` type struct? 

```go
func (d Data) MarshalJSON() ([]byte, error) 
```

and 

```go
func (d Data) UnmarshalJSON(b []byte) error
```

---

Or maybe also marshal overrides for the underlying pexv2 type(s)?

File: tbdex/offering/offering.go [hunk 3]
```diff
@@ -55,15 +59,27 @@ type OfferingPaymentMethod struct {
...
-// OfferingPayoutMethod contains all the fields from PaymentMethod, in addition to estimated settlement time.
-type OfferingPayoutMethod struct {
-	OfferingPaymentMethod
+// PayoutMethod contains all the fields from PaymentMethod, in addition to estimated settlement time.
+type PayoutMethod struct {
+	Kind                    string `json:"kind"`
+	Name                    string `json:"name,omitempty"`
+	Description             string `json:"description,omitempty"`
+	Group                   string `json:"group,omitempty"`
+	RequiredPaymentDetails  string `json:"requiredPaymentDetails,omitempty"` // TODO: change to JSON Schema type
```
Comment (KendallWeihe) [inline tbdex/offering/offering.go:68]: > // TODO: change to JSON Schema type

Are we going to add a json schema type to web5-go or is there an existing go package we'll use directly here?

File: tbdex/offering/offering.go [hunk 4]
```diff
@@ -73,7 +89,7 @@ type OfferingPayoutMethod struct {
 //   - It takes the algorithm identifier of the hash function and data to digest as input and returns
 //   - the digest of the data.
 func (o Offering) Digest() ([]byte, error) {
```
Comment (KendallWeihe) [inline tbdex/offering/offering.go:91]: I think the digest functionality here could mostly be shared at the `tbdex` root, albeit I get we want it as a receiver method for each tbdex message or resource type, maybe the receiver could call the root digest function

just a thought, not necessary here

**Review by KendallWeihe — APPROVED** [review]

File: tbdex/offering/offering.go (hunk 4 above)
Comment (mistermoe) [reply tbdex/offering/offering.go:91]: I think @phoebe-lew has that WIP

File: tbdex/offering/offering.go (hunk 3 above)
Comment (mistermoe) [reply tbdex/offering/offering.go:68]: Great Q. Undecided and open to suggestions. Would prefer something pre-existing if possible!

**Pull request merged**